hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimelineUtils.java

```diff
@@ -295,4 +296,27 @@ public static Option<HoodieInstant> getEarliestInstantForMetadataArchival(
       return Option.empty();
     }
   }
+
+  /**
+   * Find the earliest incomplete commit, deltacommit, or non-clustering replacecommit,
+   * so that the incremental pulls should be strictly before this instant.
+   * This is to guard around multi-writer scenarios where a commit starting later than
+   * another commit from a concurrent writer can finish earlier, leaving an inflight commit
+   * before a completed commit.
+   */
+  public static HoodieTimeline filterTimelineForIncrementalQueryIfNeeded(
+      HoodieTableMetaClient metaClient, HoodieTimeline completedCommitTimeline, boolean useStateTransitionTime) {
+    if (useStateTransitionTime) {
+      return completedCommitTimeline;
+    }
+    Option<HoodieInstant> firstIncompleteCommit = metaClient.getCommitsTimeline()
+        .filterInflightsAndRequested()
+        .filter(instant ->
+            !HoodieTimeline.REPLACE_COMMIT_ACTION.equals(instant.getAction())
+                || !ClusteringUtils.getClusteringPlan(metaClient, instant).isPresent())
+        .firstInstant();
+    return firstIncompleteCommit.map(
```

Comment:
Just want to clear here that it could block downstream incremental jobs if clustering job/one writer takes too much time to finish(Or failed and the owner doesn't fix it in time, it could happens especially for clustering since it usually is lower priority). Now the owner need to take more care for it.

Reply:
yes that's the intention as we need to ensure not missing data first. so it's a fallback behavior when not using state transition time